designs/0035-pin-parameters.md

```diff
@@ -0,0 +1,122 @@
+  array[1 - mu_is_data] mu_param;
+  ...
+transformed parameters {
+  real mu = mu_is data ? mu_val[1] : mu_param[1];
+  ...
+```
+
+These are inelegant and unsatisfactory, partly as they require rewriting and recompiling a model.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following section is a draft of the docs for the new `pin` keyword-value pair in `cmdstan` command-line options that would appear [here](https://mc-stan.org/docs/cmdstan-guide/command_line_options.html). We don't at this stage make any proposal for how this feature would be propagated to other Stan interfaces, they do not anticipate any difficulties. 
+
+## Command-Line Interface Overview
+
+...
+
+- `pin` - specifies values for any parameters that should be pinned, if any
+
+...
+
+### Pin model parameters argument
+
+Parameters defined in the parameters block can be 'pinned' to specific values. This is useful when debugging a model or exploring a simpler model that is nested inside an extended one.
+
+By default, no parameters are pinned. The pinned parameters are read from an input data file in JSON format using the syntax:
+```
+pin=<filepath>
+```
+The value must be a filepath to a JSON file containing pinned values for some or all of the parameters in the parameters block. 
+
+At present, there are two restrictions on parameters that can be pinned:
+ 
+- you cannot pin a subset of elements of a vector; all elements must be pinned 
+- you cannot pin constrained parameters
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+** to be discussed **
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It's another command-line argument and there are already several
+- The same thing can be achieved by re-programming the model.
+- It changes the *interpretation* of a Stan model, though in a very explicit way
+- The two restrictions, particularly cannot pin constrained parameters, might limit use-cases
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+
+I think pinning parameters at runtime is far more elegant than existing solutions. At first, I had thought about a new keyword in the Stan language itself, e.g., in a parameter constraint
+```
+parameters {
+  real<pin=0.> mu;
+}
+```
+It's certainly neater than 
+```
+data {
+  int<lower=0, upper=1> mu_is_data;
+  array[mu_is_data] mu_val;
+  ...
+parameters {
+  array[1 - mu_is_data] mu_param;
+  ...
+transformed parameters {
+  real mu = mu_is data ? mu_val[1] : mu_param[1];
+  ...
+```
+but even with `<pin=>`, pinning still requires one to change a model and recompile.
+
+# Prior art
+[prior-art]: #prior-art
+
+`PyMC` has specific functionality for pinning. See [here](https://www.pymc.io/projects/docs/en/stable/api/model/generated/pymc.model.transform.conditioning.do.html). In `PyMC`, pinning (and perhaps other similar things) are called 'interventions'. The example given in the docs is this,
+```
```

Comment:
Mark code blocks for type to turn on syntax highlighting.  For example, use
* "```stan" to start a Stan code block, and
* "```python" for Python.